Fix cargo staging for run-make tests

jieyouxu opened the pull request:
Follow-up to https://github.com/rust-lang/rust/pull/130642#issuecomment-2366891866 to make sure that when

```
$ COMPILETEST_FORCE_STAGE0=1 ./x test run-make --stage 0
```

is used, bootstrap cargo is used in order to avoid building stage 1 rustc. Note that run-make tests are usually not written with `--stage 0` in mind and some tests may rely on stage1 rustc (nightly) behavior, and it is expected that some tests will fail under this invocation.

This PR also fixes `tool::Cargo` staging in compiletest when preparing for `run-make` test mode, by chopping off a stage from the `compiler` passed to `tool::Cargo` such that when the user invokes with stage `N`

```
./x test run-make --stage N
```

the `run-make` test suite will be tested against the cargo built by stage `N` compiler. Let's take `N=1`, i.e. `--stage 1`, without chopping off a stage, previously `./x test run-make --stage 1` will cause stage 1 rustc + std to be built, then stage 2 rustc, and cargo will be produced by the stage 2 rustc, which is clearly not what we want. By chopping off a stage, it means that cargo will be produced by the stage 1 rustc.

cc #119946, #59864.
See discussions regarding the tool staging at https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/.E2.9C.94.20stage1.20run-make.20tests.20now.20need.20stage2.20rustc.20built.20for.20c.2E.2E.2E.

jieyouxu committed:
Pass bootstrap cargo when `--stage 0` and `COMPILETST_FORCE_STAGE0`

And stop passing `BOOTSTRAP_CARGO` as an env var, instead the provided
cargo should go through `--cargo-path.`

jieyouxu committed:
Fix `run-make-support` to respect per-stage cargo

jieyouxu committed:
Update `run-make` tests to use `cargo` wrapper cmd

jieyouxu commented:
I tried this locally and it didn't need to build stage1 rustc, but wanted @bjorn3 to double-check in their setup.

jieyouxu commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -1630,6 +1630,13 @@ struct Compiletest {
+/// cause many tests to fail. This env var is needed to indicate that the user understands the risk
+/// and actually intends to use `compiletest` with `--stage 0`.
+fn compiletest_force_stage0() -> bool {
+    env::var("COMPILETEST_FORCE_STAGE0").is_ok()
```
Note that this just checks the presence of `COMPILETEST_FORCE_STAGE0`, it doesn't check what value it gets.

bjorn3 replied:
I think checking if we are testing stage 0 is a better idea. The env var is what allows testing stage 0, not something that should affect the actual testing.

jieyouxu replied:
Fixed so that bootstrap cargo is only passed when `builder.top_stage == 0`.

Kobzol commented on tests/run-make/thumb-none-cortex-m/rmake.rs:
```diff
@@ -28,32 +25,25 @@ fn main() {
-    .env("RUSTFLAGS", "-Copt-level=0 -Cdebug-assertions=yes --cap-lints=warn");
-
-    cmd.run();
+    cargo
```
Shouldn't this be `cargo()`?

jieyouxu replied:
Oops yeah good catch, this doesn't run locally so I didn't catch this. I also removed the FIXME immediate above since now this uses `cargo()` cmd wrapper.

Kobzol commented:
You can r=me after fixing the `cargo` invocation.

bors commented:
:pushpin: Commit 6d8150faaeca66a34f96b2e41847e20acdf84843 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 6d8150faaeca66a34f96b2e41847e20acdf84843 -->
<!-- homu: {"type":"Approved","sha":"6d8150faaeca66a34f96b2e41847e20acdf84843","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
Failed rollup: https://github.com/rust-lang/rust/pull/130767#issuecomment-2370054534

@bors r-

Kobzol commented:
Oh, there is another thumb run-make test..

Kobzol commented on tests/run-make/thumb-none-qemu/rmake.rs:
```diff
@@ -14,49 +14,40 @@
+        .args(&["run", "--target", &target()])
+        .env("RUSTFLAGS", "-C linker=arm-none-eabi-ld -C link-arg=-Tlink.x")
+        .env("CARGO_TARGET_DIR", &target_dir)
+        .env("PATH", &path_env)
```
Now that I see it again, `PATH` and `RUSTC` should be inherited by default, right?

jieyouxu replied:
Hm, I'm not 100% sure...

Kobzol replied:
Unless we do some magic inside of run-make-support, creating a Rust `Command` should automatically propagate env. vars, if https://doc.rust-lang.org/std/process/struct.Command.html#method.env_clear is not used.

jieyouxu replied:
Actually no I think you're right, since `rmake.rs` is given `RUSTC` and `PATH` (at least on windows) which *should* mean that `cargo` as a `Command` also inherits `RUSTC` and `PATH`? I suppose it doesn't hurt to be explicit in *this* case...

Kobzol replied:
Sure, let's keep it explicit. You can r=me once CI is green.

jieyouxu replied:
Good point, I'll remove it

jieyouxu replied:
Uhh the comment timings are slightly awkward lol, I removed the unnnecessary `PATH` and `RUSTC` since they should be inherited by `rmake.rs`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11010053635/job/30571114433) [(plain)](https://github.com/rust-lang/rust/commit/c8bd9c3835e5033125893569de1552cc3ecba297/checks/30571114433/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:30ca74372d8b771363f68f939a58b017a592fae4f69398600dc51145997160f03e9652051f957840c41898984a88855e9757fa23464703a5a4ba21316ddebb04:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.59s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/compiler-builtins stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/compiler-builtins/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-18/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-18/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/compiler-builtins/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/compiler-builtins/rmake"
Testing compiler_builtins for x86_64-unknown-linux-gnu
------------------------------------------
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/run-make/compiler-builtins/rmake.rs:44:21:
thread 'main' panicked at /checkout/tests/run-make/compiler-builtins/rmake.rs:44:21:
failed to retrieve environment variable "LIB": NotPresent
------------------------------------------



```

</details>

jieyouxu commented:
Wait a minute, does stage 1 cargo rely on stage 2 rustc...?

jieyouxu committed:
Fix tool cargo being off-by-one from rustc staging

Previously if you pass compiler stage 1 to `tool::Cargo`, it will build
stage2 rustc and give you back a cargo built with stage2 rustc, which is
not what we want.

This commit adds a hack that chops off a stage from the compiler passed
to `tool::Cargo`, meaning that we will get a cargo built with stage 1
compiler, avoiding unnecessary and incorrect build of stage2 rustc and
the cargo built by that.

jieyouxu commented:
Changes since last review:

- Fixed `tool::Cargo` staging being off-by-one (see updated PR description for more details).
- Removed `PATH` and `RUSTC` since those env-vars from cargo commands invoked within `rmake.rs` should inherit the env vars.
- Reverted changes to `LIB` env var in compiler-builtins.

Kobzol commented:
Seems to work, even in combination with `download-rustc`. You can r=me once CI is green.

jieyouxu commented:
@bors r=@Kobzol rollup=iffy (some tests I can't run locally, plus bootstrap cargo staging changes for run-make tests)

bors commented:
:pushpin: Commit f5482167286424fd25266d8202e8ded6655cc9ec has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol f5482167286424fd25266d8202e8ded6655cc9ec -->
<!-- homu: {"type":"Approved","sha":"f5482167286424fd25266d8202e8ded6655cc9ec","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit f5482167286424fd25266d8202e8ded6655cc9ec with merge 3f99982c633dbca746140db60ed52ba7fa112803...
<!-- homu: {"type":"BuildStarted","head_sha":"f5482167286424fd25266d8202e8ded6655cc9ec","merge_sha":"3f99982c633dbca746140db60ed52ba7fa112803"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11023117941/job/30619081523)
Approved by: Kobzol
Pushing 3f99982c633dbca746140db60ed52ba7fa112803 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Kobzol","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11023117941/job/30619081523"},"merge_sha":"3f99982c633dbca746140db60ed52ba7fa112803"} -->

rust-timer commented:
Finished benchmarking commit (3f99982c633dbca746140db60ed52ba7fa112803): [comparison URL](https://perf.rust-lang.org/compare.html?start=363ae4188316b8b22cf6c1890bc73d84d05f70a4&end=3f99982c633dbca746140db60ed52ba7fa112803&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=363ae4188316b8b22cf6c1890bc73d84d05f70a4&end=3f99982c633dbca746140db60ed52ba7fa112803&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.1% | [2.1%, 2.1%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=363ae4188316b8b22cf6c1890bc73d84d05f70a4&end=3f99982c633dbca746140db60ed52ba7fa112803&stat=max-rss)
<details>
<summary>Results (secondary -1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.9% | [-1.9%, -1.9%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=363ae4188316b8b22cf6c1890bc73d84d05f70a4&end=3f99982c633dbca746140db60ed52ba7fa112803&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=363ae4188316b8b22cf6c1890bc73d84d05f70a4&end=3f99982c633dbca746140db60ed52ba7fa112803&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 768.432s -> 767.733s (-0.09%)
**Artifact size**: 340.84 MiB -> 340.84 MiB (0.00%)
